task2/src/main/java/backend/Teacher.java

```diff
@@ -118,27 +119,52 @@ private double[][] calculateErrorInHiddenLayers() throws IllegalAccessException
         if (neuralNetwork.isWithBias()) {
             biasErrorHidden = new double[numOfHiddenLayers][maxConnections];
         }
+        int numOfConnectionsInAllPreviousLayers = 0;
         for (int hLayer = 0; hLayer < numOfHiddenLayers; hLayer++) {
             int hiddenLayerNum = numOfHiddenLayers - hLayer;
             double[] outputs = neuralNetwork.getLayersResult()[hiddenLayerNum];
             int numOfNeuronsInPreviousLayer = neuralNetwork.getLayersResult()[hiddenLayerNum - 1].length;
+
+            int numOfNeuronsInNextLayer = 0;
+            if (hLayer == 0) {
+                numOfConnectionsInAllPreviousLayers = numOfOutputs;
+            } else {
+                numOfNeuronsInNextLayer = neuralNetwork.getLayersResult()[hiddenLayerNum + 1].length;
+            }
+            errorDerivative[hLayer + 1] = new double[numOfConnectionsInAllPreviousLayers * outputs.length];
+
+
             for (int i = 0; i < outputs.length; i++) {
                 double totalErrorDerivative = 0;
+                double neuronOutputErrorDerivative = outputs[i] * (1 - outputs[i]);
                 if (hLayer == 0) {
                     for (int j = 0; j < numOfOutputs; j++) {
                         double weight = neuralNetwork.getOutputLayer().getNeuronWeight(j, i);
-                        totalErrorDerivative += totalNeuronOutputErrorDerivativesOfOutputLayer[j] * weight;
+                        errorDerivative[hLayer + 1][j + i * numOfOutputs] = errorDerivative[0][j] * weight;
+                        totalErrorDerivative += errorDerivative[hLayer + 1][j + i * numOfOutputs];
+                    }
+                    for (int j = 0; j < numOfOutputs; j++) {
+                        errorDerivative[hLayer + 1][j + i * numOfOutputs] *= neuronOutputErrorDerivative;
                     }
                 } else {
-                    int numOfNeurons = neuralNetwork.getHiddenLayers()[hiddenLayerNum].getNeurons().length;
-                    for (int j = 0; j < numOfNeurons; j++) {
+                    for (int j = 0; j < numOfNeuronsInNextLayer; j++) {
                         double weight = neuralNetwork.getHiddenLayers()[hiddenLayerNum].getNeuronWeight(
-                                j, (i * numOfNeuronsInPreviousLayer) / numOfOutputs
+                                j, i
                         );
-                        totalErrorDerivative += totalNeuronOutputErrorDerivativesOfOutputLayer[j] * weight;
+                        for (int k = 0; k < numOfConnectionsInAllPreviousLayers / numOfNeuronsInNextLayer; k++) {
+                            int numOfDerivative = k + j * (numOfConnectionsInAllPreviousLayers / numOfNeuronsInNextLayer)
+                                    + i * numOfConnectionsInAllPreviousLayers;
+                            int numOfPreviousDerivative = j * (numOfConnectionsInAllPreviousLayers / numOfNeuronsInNextLayer) + k;
```

Comment:
This variable is used in previous too
```suggestion
                            int numOfPreviousDerivative = j * (numOfConnectionsInAllPreviousLayers / numOfNeuronsInNextLayer) + k;
                            int numOfDerivative = numOfPreviousDerivative + i * numOfConnectionsInAllPreviousLayers;
```